src/instructlab/model/phased_training.py

```diff
@@ -0,0 +1,198 @@
+# Standard
+import datetime
+import enum
+import fcntl
+import logging
+import os
+import pathlib
+import typing
+import uuid
+
+# Third Party
+import pydantic
+import rich
+import yaml
+
+logger = logging.getLogger(__name__)
+
+
+class TrainingPhases(enum.Enum):
+    TRAIN1 = "train1"
+    TRAIN2 = "train2"
+    EVAL1 = "eval1"
+    EVAL2 = "eval2"
+    DONE = "done"
+
+
+AutoDatetimeField = pydantic.Field(
+    default_factory=lambda: datetime.datetime.now(datetime.UTC)
+)
+
+
+class EvalResult(pydantic.BaseModel):
+    """Single checkpoint's final score"""
+
+    ended_at_utc: datetime.datetime = AutoDatetimeField
+    checkpoint: pydantic.DirectoryPath
+    score: float
+
+    @pydantic.field_serializer("checkpoint", "ended_at_utc")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+
+class EvalPhaseModel(pydantic.BaseModel):
+    """Stores info about evaluation phase"""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: list[pydantic.DirectoryPath]
+    finished_checkpoints: list[pydantic.DirectoryPath] = []
+    results: list[EvalResult] = []
+    best_checkpoint: EvalResult | None = None
+
+    @pydantic.field_serializer(
+        "checkpoints",
+        "finished_checkpoints",
+    )
+    def pathlibPath_list_to_str(self, paths: list[pathlib.Path]) -> list[str]:
+        return [str(path) for path in paths]
+
+    @pydantic.field_serializer("ended_at_utc")
+    def serialize_optional_datetime(self, val: datetime.datetime | None):
+        if val:
+            return str(val)
+        return val
+
+
+class TrainPhaseModel(pydantic.BaseModel):
+    """Stores info about training phase. Doesn't store training args because user can get those elsewhere, e.g. config."""
+
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    checkpoints: pydantic.DirectoryPath
+    # TODO: might want training args here, but can't currently do it
+    # without side-effects because model instance is changed inside of _training_phase
+
+    @pydantic.field_serializer("checkpoints")
+    def call_str_constructor(self, val: typing.Any) -> str:
+        return str(val)
+
+    @pydantic.field_serializer("ended_at_utc")
+    def serialize_optional_datetime(self, val: datetime.datetime | None):
+        if val:
+            return str(val)
+        return val
+
+
+class JournalModel(pydantic.BaseModel):
+    run_id: uuid.UUID = pydantic.Field(default_factory=uuid.uuid4)
+    started_at_utc: datetime.datetime = AutoDatetimeField
+    ended_at_utc: datetime.datetime | None = None
+    current_phase: TrainingPhases = TrainingPhases.TRAIN1
+    train_1: TrainPhaseModel | None = None
```

Comment:
The only reason I didn't do that was because the different phases _could_ have other meanings in the future. We might decide that phase_1 is actually called "uptake" while phase_2 is called "practice" or something. Therefore I wanted to separate the implementation from the scientific reasoning- there's a first, and then a second phase.

Reply:
https://github.com/instructlab/instructlab/pull/2236#discussion_r1777665328